File: Data_Ingestion/data_ingestion_app/core_modules.py
"""
Core modules
"""
import requests
import pymongo
import os
import datetime
MONGODB_USERNAME = os.environ.get("MONGODB_USERNAME")
MONGODB_PASSWORD = os.environ.get("MONGODB_PASSWORD")
MONGODB_PORT = os.environ.get("MONGODB_PORT")
MONGODB_DATABASE = os.environ.get("MONGODB_DATABASE")
MONGODB_COLLECTION = os.environ.get("MONGODB_COLLECTION")
MONGODB_SERVER = os.environ.get("MONGODB_SERVER")
TIMEOUT = 1  # seconds

def _return_database_connection():
    """
    Internal function to return a collection obj
    :return:
    """

    client = pymongo.MongoClient(f"mongodb://{MONGODB_SERVER}:{MONGODB_PORT}/", username=MONGODB_USERNAME, password=MONGODB_PASSWORD)
    db = client["DataIngestionDb"]
    col = db["Data_Collection"]
    return client, col

def collect_data(data_source, timeout=TIMEOUT):
    """
    Collects data from data source
    :param data_source: url of data source
    :return: response_code, data
    """
    response_code = 200
    data = None
    try:
        data  = requests.get(data_source, timeout=timeout)
        data = data.json()
    except requests.exceptions.Timeout:
        print("Request timed out")
        response_code = 408 # timeout
    except Exception as e:
        response_code = 500 # internal server error
        print("Exception occurred while collecting data", e)

    return response_code, data

def transform_data(input_data, data_source):
    """
    Transforms given data
    :param input_data:
    :param data_source:
    :return:
    """
    response_code = 200
    output_data = {}
    try:
        output_data = {"data": input_data, "ingested_at": datetime.datetime.utcnow(), "source": data_source}
    except Exception as e:
        response_code = 500
        print("Exception occurred while transforming data", e)

    return response_code, output_data

def store_data(data):
    """
    Stores given data
    :param data: valid data
    :return:
    """
    response_code = 200
    try:
        client, collection_obj = _return_database_connection()
        result = collection_obj.insert_one(data)
        print(f"Inserted document ID: {result.inserted_id}")
        client.close()
    except Exception as e:
        response_code = 500
        print("Exception occurred while storing data", e)

    return response_code

def fetch_inserted_data():
    """
    Fetch last inserted data
    Note: Assuming latest data fetch is required. Requirements are not specific
    :return:
    """
    response_code = 200
    data = None
    try:
        client, collection_obj = _return_database_connection()
        data = collection_obj.find_one(sort=[('ingested_at', pymongo.DESCENDING)])
        client.close()
    except Exception as e:
        response_code = 500
        print("Exception occurred while fetching data", e)

    return response_code, data

File: Data_Ingestion/data_ingestion_app/core_modules.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "net/http"
    "os"
    "time"
    "go.mongodb.org/mongo-driver/bson"
    "go.mongodb.org/mongo-driver/mongo"
    "go.mongodb.org/mongo-driver/mongo/options"
)

var (
    mongoUser     = os.Getenv("MONGODB_USERNAME")
    mongoPassword = os.Getenv("MONGODB_PASSWORD")
    mongoPort     = os.Getenv("MONGODB_PORT")
    mongoDB       = os.Getenv("MONGODB_DATABASE")
    mongoColl     = os.Getenv("MONGODB_COLLECTION")
    mongoServer   = os.Getenv("MONGODB_SERVER")
    timeout       = 1 * time.Second
)

func getMongoCollection() (*mongo.Client, *mongo.Collection, error) {
    uri := fmt.Sprintf("mongodb://%s:%s", mongoServer, mongoPort)
    clientOpts := options.Client().ApplyURI(uri).SetAuth(options.Credential{
        Username: mongoUser,
        Password: mongoPassword,
    })
    client, err := mongo.Connect(context.TODO(), clientOpts)
    if err != nil {
        return nil, nil, err
    }
    collection := client.Database("DataIngestionDb").Collection("Data_Collection")
    return client, collection, nil
}

func collectData(dataSource string) (int, map[string]interface{}) {
    client := http.Client{Timeout: timeout}
    resp, err := client.Get(dataSource)
    if err != nil {
        fmt.Println("Request error:", err)
        return http.StatusRequestTimeout, nil
    }
    defer resp.Body.Close()

    var data map[string]interface{}
    json.NewDecoder(resp.Body).Decode(data)
    fmt.Println(data)

    return http.StatusOK, data
}

func transformData(input map[string]interface{}, source string) (int, map[string]interface{}) {
    output := map[string]interface{}{
        "data":       input,
        "ingested_at": time.Now().UTC(),
        "source":     source,
    }
    return http.StatusOK, output
}

func storeData(data map[string]interface{}) int {
    client, collection, err := getMongoCollection()
    if err != nil {
        fmt.Println("Mongo connection error:", err)
        return http.StatusInternalServerError
    }
    defer client.Disconnect(context.TODO())
    res, err := collection.InsertOne(context.TODO(), data)
    if err != nil {
        fmt.Println("Insert error:", err)
        return http.StatusInternalServerError
    }
    fmt.Println("Inserted ID:", res.InsertedID)
    return http.StatusOK
}

func fetchInsertedData() (int, map[string]interface{}) {
    client, collection, err := getMongoCollection()
    if err != nil {
        fmt.Println("Mongo connection error:", err)
        return http.StatusInternalServerError, nil
    }
    defer client.Disconnect(context.TODO())
    var result map[string]interface{}
    err = collection.FindOne(context.TODO(), bson.M{}, options.FindOne().SetSort(bson.D{{"ingested_at", -1}})).Decode(&result)
    if err != nil {
        fmt.Println("Fetch error:", err)
        return http.StatusInternalServerError, nil
    }
    return http.StatusOK, result
}



